cmd/main.go

```diff
@@ -287,23 +332,52 @@ func main() {
 	}
 }
 
-type fieldIndexer struct {
-	mgr ctrl.Manager
+// restConfigForLogicalClusterHostingAPIExport returns a *rest.Config properly configured
+// to communicate with the endpoint for the APIExport's virtual workspace.
+func restConfigForLogicalClusterHostingAPIExport(
+	ctx context.Context, cfg *rest.Config, apiExportName string) (*rest.Config, error) {
+	apiExportClient, err := client.New(cfg, client.Options{
+		Scheme: controlplanev1alpha1.Scheme,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("error creating APIExport client: %w", err)
+	}
+
+	apiExport := &apisv1alpha1.APIExport{}
+	if err := apiExportClient.Get(ctx, types.NamespacedName{Name: apiExportName}, apiExport); err != nil {
+		return nil, fmt.Errorf("error getting APIExport %q: %w", apiExportName, err)
+	}
+	// This field is deprecated, but the alternative is not feasible for us.
+	// We do not use Partitions. Without partitions, the controller does not populate the endpoints.
+	if len(apiExport.Status.VirtualWorkspaces) < 1 { // nolint:staticcheck
+		return nil, fmt.Errorf("APIExport %q status.virtualWorkspaces is empty", apiExportName)
+	}
+
+	// create a new rest.Config with the APIExport's virtual workspace URL
+	exportConfig := rest.CopyConfig(cfg)
+	exportConfig.Host = apiExport.Status.VirtualWorkspaces[0].URL // nolint:staticcheck
+
+	return exportConfig, nil
 }
 
-func (f *fieldIndexer) Start(ctx context.Context) error {
-	return f.mgr.GetFieldIndexer().IndexField(
-		ctx,
-		&controlplanev1alpha1.GardenerShootControlPlane{},
-		controlplanev1alpha1.ShootReferenceIndexKey,
-		func(obj client.Object) []string {
-			controlPlane, ok := obj.(*controlplanev1alpha1.GardenerShootControlPlane)
-			if !ok {
-				return nil
-			}
-			return []string{client.ObjectKey{
-				Namespace: controlPlane.Spec.ProjectNamespace,
-				Name:      controlPlane.GetName()}.String(),
+func kcpAPIsGroupPresent(cfg *rest.Config) (bool, error) {
```

Comment:
small suggested renaming(?):

```suggestion
func hasKcpAPIGroups(cfg *rest.Config) (bool, error) {
```

Reply:
This is from the official example of `kcp`.
I am in favor of your renaming though. 😅

On a side note: I am actually not sure why this is not some helper function in the `kcp` packages, as I feel like all, or at least a significant amount of, `kcp`-aware controllers need to check for something like this.
Maybe I create an issue to ask for something like this.